ipn/ipnlocal/local.go

```diff
@@ -140,9 +141,23 @@ type LocalBackend struct {
 	statusChanged *sync.Cond
 }
 
+type clientGen func(controlclient.Options) (controlclient.Client, error)
```

Comment:
Rather than pass in a constructor function (which I've found ends up making the dependencies harder to follow), can the caller construct and pass in a controlclient.Client? That should still be very friendly to tests, and only requires a modest bit of extra work out in/near func main (which is already constructing and wiring up many pieces)

Reply:
Unfortunately not without significant surgery. ipn.Backend is defined to do something appropriate if you re-run Start() over and over (the idea being that there has to be *some* code that works correctly if you keep connecting to it), but controlclient.AutoClient doesn't work that way; you are supposed to shut it down and make a new one, which is what ipn.Backend does when we reset the connection.

We could change that behaviour of controlclient.AutoClient (since, after all, it only really has that one caller). But generally speaking I really don't like objects that get recycled. It's really error prone. ipn.Backend was supposed to be a tiny lightweight interconnection between heavier objects (that it could create and destroy as needed), but it has, uh, grown.

Reply:
I see. Well, as someone who removed a lot of these generator functions because they obscured the control flow a huge amount, I may try to get that object construction/destruction encapsulated into controlclient, so that ipn can have a long-lived handle to play with, but that's definitely not something for this refactor.